packages/next/src/client/app-index.tsx

```diff
@@ -33,6 +44,72 @@ window.addEventListener('error', (ev: WindowEventMap['error']): void => {
 
 const appElement: HTMLElement | Document | null = document
 
+function finishRunningAction(
+  actionQueue: AppRouterActionQueue,
+  setState: DispatchStatePromise
+) {
+  if (actionQueue.pending !== null) {
+    actionQueue.pending = actionQueue.pending.next
+    if (actionQueue.pending !== null) {
+      // eslint-disable-next-line @typescript-eslint/no-use-before-define
+      runAction(actionQueue, actionQueue.pending.payload, setState)
+    }
+  }
+}
+async function runAction(
+  actionQueue: AppRouterActionQueue,
+  payload: ReducerActions,
+  setState: DispatchStatePromise
+) {
+  const prevState = actionQueue.state
+  const promise = actionQueue.action(prevState, payload)
+
+  promise.then(
+    (nextState: AppRouterState | null) => {
+      actionQueue.state = nextState
+
+      if (actionQueue.devToolsInstance) {
+        actionQueue.devToolsInstance.send(payload, nextState)
+      }
+
+      finishRunningAction(actionQueue, setState)
+    },
+    () => {
+      finishRunningAction(actionQueue, setState)
+    }
+  )
+
+  startTransition(() => {
+    setState(promise)
```

Comment:
This setState needs to move into `dispatchAction` so that it runs immediately even if it's blocked by earlier actions. The way transitions work in React is that we batch all transition updates that occur within the same event:

```js
function onClick() {
  navigate(...);
  startTransition(() => setSomeOtherState());
}
```

In order for React to know that these two updates are related, they both need to call `setState` before the `onClick` handler ends.

 (The code snippet I sent you of `useFormState` in React doesn't do this because it takes advantage of the fact that async actions are batched but for conceptual simplicity it probably should, too.)

You can do this by creating a new promise object, saving its `resolve` method, calling setState with that, and then later calling `resolve`.

When we later implement batching, we can be clever and reuse the same promise object for multiple actions.